cmd/scollector/collectors/google_cloudbilling.go

```diff
@@ -0,0 +1,234 @@
+
+	"encoding/base64"
+
+	"encoding/json"
+
+	"strings"
+
+	"math"
+
+	"bosun.org/cmd/scollector/conf"
+	"bosun.org/metadata"
+	"bosun.org/opentsdb"
+	"bosun.org/slog"
+)
+
+var gcbConf = conf.GoogleCloudBilling{}
+
+const (
+	gcbHoursInDay = 24
+	gcbUsageDesc  = "Usage of consuming given Google Cloud services"
+	gcbCostDesc   = "Cost of consuming given Google Cloud services"
+)
+
+// GoogleCloudBilling instantiates the GoogleCloudBilling collector
+func GoogleCloudBilling(bucketName, authBase64 string) error {
+	if bucketName != "" {
+		gcbConf = conf.GoogleCloudBilling{
+			BucketName: bucketName,
+			AuthBase64: authBase64,
+		}
+		collectors = append(collectors, &IntervalCollector{
+			F:        c_googlecloudbilling,
+			Interval: 1 * time.Hour,
+		})
+	}
+
+	return nil
+}
+
+func c_googlecloudbilling() (opentsdb.MultiDataPoint, error) {
+	var md opentsdb.MultiDataPoint
+	//Create a connection to the Google Cloud services
+	gcClient, err := createGCStorageClient()
+	if err != nil {
+		return nil, err
+	}
+	//Get the list of available files from the storage bucket
+	reports, err := gcClient.Objects.List(gcbConf.BucketName).Do()
+	if err != nil {
+		return nil, err
+	}
+	//We want to process bills from the last 4 days. This is because Google Cloud seems to have significant
+	//lag when generating the bill exports. They can be 2 or more days old by the time the bill is put into the
+	//storage bucket.
+	dateStart := time.Now().Add(-96 * time.Hour)
+	dateRangeStart := time.Date(dateStart.Year(), dateStart.Month(), dateStart.Day(), 0, 0, 0, 0, time.UTC)
+	dateRangeEnd := dateRangeStart.Add(120 * time.Hour).Add(-1 * time.Second)
+	dateLayout := "2006-01-02"
+	//Run through each item in the bucket
+	for _, item := range reports.Items {
+		//Is it a JSON file?
+		fileNameParts := strings.Split(item.Name, ".")
+		fileExt := strings.ToLower(fileNameParts[len(fileNameParts)-1])
+		if fileExt != "json" {
+			continue
+		}
+		//Check the date. We only want the _latest_ file. In case there are extra . in the filename,
+		//the date will always be immediately before the extension
+		fileName := fileNameParts[len(fileNameParts)-2]
+		//If the last part of the filename does not have 10 or more characters, then it's not datestamped
+		if len(fileName) < 10 {
+			continue
+		}
+		dateString := fileName[len(fileName)-10:]
+		dateDate, err := time.Parse(dateLayout, dateString)
+		if err != nil {
+			continue
+		}
+		if !(dateDate.Before(dateRangeEnd) && dateDate.After(dateRangeStart)) {
+			continue
+		}
+		//Ok, so we want to process this. Now we need to download the file, convert it to a string containing the JSON,
+		//unmarshall that into the bill struct and send it off for processing
+		thisItemResponse, err := gcClient.Objects.Get(item.Bucket, item.Name).Download()
+		if err != nil {
+			return md, fmt.Errorf("gcloud: Could not download %s: %s", item.Name, err)
+		}
+		buf := bytes.NewBuffer(make([]byte, 0, thisItemResponse.ContentLength))
```

Comment:
just make a json.NewDecoder(thisItemResponse.Body) and decode from that directly. No need to read the content.